s3prl/downstream/speech_translation/expert.py

```diff
@@ -0,0 +1,506 @@
+
+            logger:
+                Tensorboard SummaryWriter
+                please use f'{your_task_name}/{mode}-{key}' as key name to log your contents,
+                preventing conflict with the logging of other tasks
+
+            global_step:
+                The global_step when training, which is helpful for Tensorboard logging
+
+            batch_ids:
+                The batches contained in records when enumerating over the dataloader
+
+            total_batch_num:
+                The total amount of batches in the dataloader
+        
+        Return:
+            a list of string
+                Each string is a filename we wish to use to save the current model
+                according to the evaluation result, like the best.ckpt on the dev set
+                You can return nothing or an empty list when no need to save the checkpoint
+        """
+        save_names = []
+
+        if mode in ['train', 'dev']:
+
+            ave_loss = sum(records['loss'])/len(records['loss'])
+            logger.add_scalar(
+                f'st/{mode}-loss',
+                ave_loss,
+                global_step=global_step
+            )
+
+            if self.use_asr:
+
+                ave_st_loss = sum(records['st_loss'])/len(records['st_loss'])
+                logger.add_scalar(
+                    f'st/{mode}-st_loss',
+                    ave_st_loss,
+                    global_step=global_step
+                )
+
+                ave_asr_loss = sum(records['asr_loss'])/len(records['asr_loss'])
+                logger.add_scalar(
+                    f'asr/{mode}-asr_loss',
+                    ave_asr_loss,
+                    global_step=global_step
+                )
+
+        if mode in ['dev', 'test']:
+
+            bleu = self._metric(records['hyps'], records['refs'])
+            logger.add_scalar(
+                f'st/{mode}-bleu',
+                bleu.score,
+                global_step=global_step
+            )
+            for i in range(4):
+                logger.add_scalar(
+                    f'st/{mode}-bleu{i+1}',
+                    bleu.precisions[i],
+                    global_step=global_step
+                )
+
+            if bleu.score > self.best_score and mode == 'dev':
+                self.best_score = torch.ones(1) * bleu.score
+                save_names.append(f'{mode}-best.ckpt') 
+            
+            with open(f'{self.output_prefix}-st-{mode}.tsv', 'w') as f:
+                print('id', 'hyp', 'ref', sep='\t', file=f)
+                for idx, hyp, ref in zip(records['ids'], records['hyps'], records['refs']):
+                    print(idx, hyp, ref, sep='\t', file=f)
+
+            for i in range(min(len(records['ids']), 5)):
+                tqdm.write(f"[st] [id]: {records['ids'][i]}, [hyp]: {records['hyps'][i]}, [ref]: {records['refs'][i]}")
+
+            tqdm.write(f'[BLEU]: {bleu.score}')
+            print(bleu)
```

Comment:
Yes, these lines are for debugging which show few samples for user to quickly check the decoding results and also the score. Since we already have output files and tensorboard to tract the result and performance, I think we could just remove them.